<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Using the dupRadar package • dupRadar</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Using the dupRadar package">
<meta property="og:description" content="dupRadar">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">dupRadar</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.20.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/dupRadar.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><link href="dupRadar_files/anchor-sections-1.0/anchor-sections.css" rel="stylesheet">
<script src="dupRadar_files/anchor-sections-1.0/anchor-sections.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Using the dupRadar package</h1>
                        <h4 class="author">Sergi Sayols, Holger Klein</h4>
            
            <h4 class="date">June 23, 2015</h4>
      
      
      <div class="hidden name"><code>dupRadar.Rmd</code></div>

    </div>

    
    
<!--
% \VignetteEngine{knitr::rmarkdown}
% \VignetteIndexEntry{Using dupRadar}
-->
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script><script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script><div id="introduction-to-dupradar" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction-to-dupradar" class="anchor"></a>Introduction to dupRadar</h1>
<p>RNA-Seq experiments are a common strategy nowadays to quantify the gene expression levels in cells. Due to its higher sensitivity compared to arrays and the ability to discover novel features, makes them the choice for most modern experiments.</p>
<p>In RNA-Seq - as in all other NGS applications - quality control is essential. Current NGS workflows usually involve PCR steps at some point, which involves the danger of PCR artefacts and over-amplification of reads. For common DNA-based assays PCR duplicates are often simply removed before further analysis and their overall fraction or the read multiplicity taken as quality metrics. In RNA-Seq however, apart from the technical PCR duplicates, there is a strong source for biological read duplicates: for a given gene length and sequencing depth there exists an expression level beyond which it is not possible to add place more reads inside a gene locus without placing a second read exactly on the position of an already existing read. For this reason the overall duplication rate is not a useful measure for RNA-Seq.</p>
<p>As in the NGS/RNA-Seq QC ecosystem there were no suitable tools to address this question, we set out to develop a new tool. The <em>dupRadar</em> package gives an insight into the duplication problem by graphically relating the gene expression level and the duplication rate present on it. Thus, failed experiments can be easily identified at a glance.</p>
<p><strong>Note: by now RNA-Seq protocols have matured so that for bulk RNA protocols data rarely suffer from technical duplicates. With newer low-input or single cell RNA-Seq protocols technical duplicates possible problems are worth to be checked for by default, especially if protocols are pushed to or beyond their boundaries. Paired-end libraries make the distinction between duplicates due to highly expressed genes and PCR duplicates easier, but the problem itself is not completely solved, especially at higher sequencing depths.</strong></p>
</div>
<div id="getting-started-using-dupradar" class="section level1">
<h1 class="hasAnchor">
<a href="#getting-started-using-dupradar" class="anchor"></a>Getting started using dupRadar</h1>
<div id="preparing-your-data" class="section level2">
<h2 class="hasAnchor">
<a href="#preparing-your-data" class="anchor"></a>Preparing your data</h2>
<p>Previous to running the duplication rate analysis, the BAM file with your mapped reads has to be duplicate marked with a like Picard, or the faster BamUtil dedup BioBamBam. The <em>dupRadar</em> package only works with duplicate marked BAM files.</p>
<p>If you do not have/want a duplication marking step in your default pipeline, the <em>dupRadar</em> package includes, for your convenience, wrappers to properly call some of these tools from your R session. Note that you still have to supply the path of the dupmarker installation though:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://www.bioconductor.org/packages/dupRadar">dupRadar</a></span><span class="op">)</span></code></pre></div>
<p>Now, simply call the wrapper:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># call the duplicate marker and analyze the reads</span>
<span class="va">bamDuprm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/markDuplicates.html">markDuplicates</a></span><span class="op">(</span>dupremover<span class="op">=</span><span class="st">"bamutil"</span>,
               bam<span class="op">=</span><span class="st">"test.bam"</span>,
               path<span class="op">=</span><span class="st">"/opt/bamUtil-master/bin"</span>,
               rminput<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<p>Simply specify which tool to use, the path where this tool is installed, and the input bam file to be analyzed. After marking duplicates, it’s safe to remove to original BAM file in order to save space.</p>
<p>The <em>dupRadar</em> package currently comes with support for:</p>
<ul>
<li><a href="http://broadinstitute.github.io/picard">Picard MarkDuplicates</a></li>
<li><a href="http://genome.sph.umich.edu/wiki/BamUtil">BamUtil</a></li>
</ul>
<p>After the BAM file is marked for duplicates, dupRadar is ready to analyze how the duplication rate is related with the estimated gene expression levels.</p>
</div>
<div id="a-gtf-file" class="section level2">
<h2 class="hasAnchor">
<a href="#a-gtf-file" class="anchor"></a>A GTF file</h2>
<p>Unless there is any specific reason, dupRadar can use the same GTF file that will be later used to count the reads falling on features.</p>
<p>A valid GTF file can be obtained from UCSC, Ensembl, the iGenomes or other projects.</p>
<p>Note that the resulting duplication rate plots depend on the GTF annotation used. GTF files from the gencode projects result in a less clear picture of duplication rates, as there are many more features and feature types annotated, which overlap heavily as well. In some cases creating the plots only using subsets of gencode annotation files (e.g. just protein coding genes) serve the QC purpose of this tool better.</p>
</div>
<div id="annotationhub-as-a-source-of-gtf-files" class="section level2">
<h2 class="hasAnchor">
<a href="#annotationhub-as-a-source-of-gtf-files" class="anchor"></a>AnnotationHub as a source of GTF files</h2>
<p>The Bioconductor <em>AnnotationHub</em> package provides an alternative approach to obtain annotation in GTF format from entities such as Ensembl, UCSC, ENCODE, and 1000 Genomes projects.</p>
<p>This is partly outlined in the AnnotationHub ‘HOWTO’ vignette section “Ensembl GTF and FASTA files for TxDb gene models and sequence queries”; for the Takifugu example, the downloaded GTF file is available from the cache.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/warning.html">suppressWarnings</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span><span class="op">(</span><span class="va">AnnotationHub</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
    <span class="va">ah</span> <span class="op">=</span> <span class="fu">AnnotationHub</span><span class="op">(</span><span class="op">)</span>
    <span class="fu">query</span><span class="op">(</span><span class="va">ah</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"ensembl"</span>, <span class="st">"80"</span>, <span class="st">"Takifugu"</span>, <span class="st">"gtf"</span><span class="op">)</span><span class="op">)</span>    <span class="co"># discovery</span>
    <span class="fu">cache</span><span class="op">(</span><span class="va">ah</span><span class="op">[</span><span class="st">"AH47101"</span><span class="op">]</span><span class="op">)</span>                                <span class="co"># retrieve / file path</span>
<span class="op">}</span></code></pre></div>
</div>
</div>
<div id="duprate-demo-data" class="section level1">
<h1 class="hasAnchor">
<a href="#duprate-demo-data" class="anchor"></a>dupRate demo data</h1>
<p>In the package we include two precomputed duplication matrices for two RNASeq experiments used as examples of a good and a failed (in terms of high redundancy of reads) experiments. The experiments come from the ENCODE project, as a source of a wide variety of protocols, library types and sequencing facilities.</p>
<p>Load the example dataset with:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/attach.html">attach</a></span><span class="op">(</span><span class="va">dupRadar_examples</span><span class="op">)</span></code></pre></div>
</div>
<div id="the-duplication-rate-analysis" class="section level1">
<h1 class="hasAnchor">
<a href="#the-duplication-rate-analysis" class="anchor"></a>The duplication rate analysis</h1>
<p>The analysis requires some info about the sequencing run and library preparation protocol:</p>
<ul>
<li>The strandess information about the reads: is the sequencing strand specific? if so, are the reads reversely sequenced?</li>
<li>Are the reads paired, or single?</li>
</ul>
<p>Due to its phenomenal performance, internally we use the featureCounts() function from the Bioconductor <em><a href="https://bioconductor.org/packages/3.11/Rsubread">Rsubread</a></em> package, which also supports multiple threads.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># The call parameters: </span>
<span class="va">bamDuprm</span> <span class="op">&lt;-</span> <span class="st">"test_duprm.bam"</span>    <span class="co"># the duplicate marked bam file</span>
<span class="va">gtf</span> <span class="op">&lt;-</span> <span class="st">"genes.gtf"</span>  <span class="co"># the gene model</span>
<span class="va">stranded</span> <span class="op">&lt;-</span> <span class="fl">2</span>       <span class="co"># '0' (unstranded), '1' (stranded) and '2' (reversely stranded)</span>
<span class="va">paired</span>   <span class="op">&lt;-</span> <span class="cn">FALSE</span>   <span class="co"># is the library paired end?</span>
<span class="va">threads</span>  <span class="op">&lt;-</span> <span class="fl">4</span>       <span class="co"># number of threads to be used</span>

<span class="co"># Duplication rate analysis</span>
<span class="va">dm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/analyzeDuprates.html">analyzeDuprates</a></span><span class="op">(</span><span class="va">bamDuprm</span>,<span class="va">gtf</span>,<span class="va">stranded</span>,<span class="va">paired</span>,<span class="va">threads</span><span class="op">)</span></code></pre></div>
<p>The duplication matrix contains read counts in different scenarios and RPK and RPKM values for every gene.</p>
<table class="table">
<colgroup>
<col width="4%">
<col width="6%">
<col width="8%">
<col width="11%">
<col width="7%">
<col width="8%">
<col width="7%">
<col width="6%">
<col width="5%">
<col width="8%">
<col width="5%">
<col width="5%">
<col width="6%">
<col width="6%">
</colgroup>
<thead><tr class="header">
<th align="left">ID</th>
<th align="right">geneLength</th>
<th align="right">allCountsMulti</th>
<th align="right">filteredCountsMulti</th>
<th align="right">dupRateMulti</th>
<th align="right">dupsPerIdMulti</th>
<th align="right">RPKMulti</th>
<th align="right">RPKMMulti</th>
<th align="right">allCounts</th>
<th align="right">filteredCounts</th>
<th align="right">dupRate</th>
<th align="right">dupsPerId</th>
<th align="right">RPK</th>
<th align="right">RPKM</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Xkr4</td>
<td align="right">3634</td>
<td align="right">2</td>
<td align="right">2</td>
<td align="right">0.0000000</td>
<td align="right">0</td>
<td align="right">0.5503577</td>
<td align="right">0.0117159</td>
<td align="right">2</td>
<td align="right">2</td>
<td align="right">0.0000000</td>
<td align="right">0</td>
<td align="right">0.5503577</td>
<td align="right">0.0117159</td>
</tr>
<tr class="even">
<td align="left">Rp1</td>
<td align="right">9747</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">NaN</td>
<td align="right">0</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">NaN</td>
<td align="right">0</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
</tr>
<tr class="odd">
<td align="left">Sox17</td>
<td align="right">3130</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">NaN</td>
<td align="right">0</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">NaN</td>
<td align="right">0</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
</tr>
<tr class="even">
<td align="left">Mrpl15</td>
<td align="right">4203</td>
<td align="right">429</td>
<td align="right">261</td>
<td align="right">0.3916084</td>
<td align="right">168</td>
<td align="right">102.0699500</td>
<td align="right">2.1728436</td>
<td align="right">419</td>
<td align="right">258</td>
<td align="right">0.3842482</td>
<td align="right">161</td>
<td align="right">99.6906971</td>
<td align="right">2.1221946</td>
</tr>
<tr class="odd">
<td align="left">Lypla1</td>
<td align="right">2433</td>
<td align="right">1106</td>
<td align="right">576</td>
<td align="right">0.4792043</td>
<td align="right">530</td>
<td align="right">454.5828196</td>
<td align="right">9.6770634</td>
<td align="right">1069</td>
<td align="right">562</td>
<td align="right">0.4742750</td>
<td align="right">507</td>
<td align="right">439.3752569</td>
<td align="right">9.3533280</td>
</tr>
<tr class="even">
<td align="left">Tcea1</td>
<td align="right">2847</td>
<td align="right">2890</td>
<td align="right">1046</td>
<td align="right">0.6380623</td>
<td align="right">1844</td>
<td align="right">1015.1036178</td>
<td align="right">21.6093121</td>
<td align="right">1822</td>
<td align="right">696</td>
<td align="right">0.6180022</td>
<td align="right">1126</td>
<td align="right">639.9719002</td>
<td align="right">13.6235871</td>
</tr>
</tbody>
</table>
</div>
<div id="plotting-and-interpretation" class="section level1">
<h1 class="hasAnchor">
<a href="#plotting-and-interpretation" class="anchor"></a>Plotting and interpretation</h1>
<p>The number of reads per base assigned to a gene in an ideal RNA-Seq data set is expected to be proportional to the abundance of its transcripts in the sample. For lowly expressed genes we expect read duplication to happen rarely by chance, while for highly expressed genes - depending on the total sequencing depth - we expect read duplication to happen often.</p>
<p>A good way to learn if a dataset is following this trend is by relating the normalized number of counts per gene (RPK, as a quantification of the gene expression) and the fraction represented by duplicated reads.</p>
<p>The <em>dupRadar</em> offers several functions to assess this relationship. The most prominent may be the 2d density scatter plot:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## make a duprate plot (blue cloud)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="../reference/duprateExpDensPlot.html">duprateExpDensPlot</a></span><span class="op">(</span>DupMat<span class="op">=</span><span class="va">dm</span><span class="op">)</span>       <span class="co"># a good looking plot</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/title.html">title</a></span><span class="op">(</span><span class="st">"good experiment"</span><span class="op">)</span>
<span class="fu"><a href="../reference/duprateExpDensPlot.html">duprateExpDensPlot</a></span><span class="op">(</span>DupMat<span class="op">=</span><span class="va">dm.bad</span><span class="op">)</span>   <span class="co"># a dataset with duplication problems</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/title.html">title</a></span><span class="op">(</span><span class="st">"duplication problems"</span><span class="op">)</span></code></pre></div>
<p><img src="dupRadar_files/figure-html/unnamed-chunk-5-1.png" width="1344"></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## duprate boxplot</span>
<span class="fu"><a href="../reference/duprateExpBoxplot.html">duprateExpBoxplot</a></span><span class="op">(</span>DupMat<span class="op">=</span><span class="va">dm</span><span class="op">)</span>        <span class="co"># a good looking plot</span>
<span class="fu"><a href="../reference/duprateExpBoxplot.html">duprateExpBoxplot</a></span><span class="op">(</span>DupMat<span class="op">=</span><span class="va">dm.bad</span><span class="op">)</span>    <span class="co"># a dataset with duplication problems</span></code></pre></div>
<p><img src="dupRadar_files/figure-html/unnamed-chunk-5-2.png" width="1344"></p>
<p>The <em>duprateExpDensPlot</em> has helper lines at the threshold of 1 read/bp and at 0.5 RPKM. Moreover by default a generalized linear model is fit to the data and overplotted (see also below).</p>
<div id="fitting-a-model-into-the-data" class="section level2">
<h2 class="hasAnchor">
<a href="#fitting-a-model-into-the-data" class="anchor"></a>Fitting a model into the data</h2>
<p>To summarize the relationship between duplication rates and gene expression, we propose fitting a logistic regression curve onto the data. With the coefficients of the fitted model, one can get an idea of the initial duplication rate at low read counts (Intercept) and the progression of the duplication rate along with the progression of the read counts (Slope).</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/duprateExpDensPlot.html">duprateExpDensPlot</a></span><span class="op">(</span>DupMat<span class="op">=</span><span class="va">dm</span><span class="op">)</span></code></pre></div>
<p><img src="dupRadar_files/figure-html/unnamed-chunk-6-1.png" width="672"></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># or, just to get the fitted model without plot</span>
<span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/duprateExpFit.html">duprateExpFit</a></span><span class="op">(</span>DupMat<span class="op">=</span><span class="va">dm</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="st">"duprate at low read counts: "</span>,<span class="va">fit</span><span class="op">$</span><span class="va">intercept</span>,<span class="st">"\n"</span>,
    <span class="st">"progression of the duplication rate: "</span>,<span class="va">fit</span><span class="op">$</span><span class="va">slope</span>,fill<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<pre><code>## duprate at low read counts:  0.04075061 
##  progression of the duplication rate:  
## 3.186793</code></pre>
<p>Our main use case for that function is the condensation of the plots into quality metrics that can be used for automatic flagging of possibly problematic samples in large experiments or aggregation with other quality metrics in large tables to analyse interdependencies.</p>
<p>The <em>duprateExpBoxplot</em> plot shows the range of the duplication rates at 5% bins (default) along the distribution of RPK gene counts. The x-axis displays the quantile of the RPK distribution, and the average RPK of the genes contained in this quantile.</p>
<p>Individual genes can be identified in the plot:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## INTERACTIVE: identify single points on screen (name="ID" column of dm)</span>
<span class="fu"><a href="../reference/duprateExpPlot.html">duprateExpPlot</a></span><span class="op">(</span>DupMat<span class="op">=</span><span class="va">dm</span><span class="op">)</span>       <span class="co"># a good looking plot</span>
<span class="fu"><a href="../reference/duprateExpIdentify.html">duprateExpIdentify</a></span><span class="op">(</span>DupMat<span class="op">=</span><span class="va">dm</span><span class="op">)</span></code></pre></div>
<p>One can also call the function <em>duprateExpPlot</em> to get smooth color density representation of the same data:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span>
<span class="va">cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/colorRamp.html">colorRampPalette</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"black"</span>,<span class="st">"blue"</span>,<span class="st">"green"</span>,<span class="st">"yellow"</span>,<span class="st">"red"</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="../reference/duprateExpPlot.html">duprateExpPlot</a></span><span class="op">(</span>DupMat<span class="op">=</span><span class="va">dm</span>,addLegend<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>
<span class="fu"><a href="../reference/duprateExpPlot.html">duprateExpPlot</a></span><span class="op">(</span>DupMat<span class="op">=</span><span class="va">dm.bad</span>,addLegend<span class="op">=</span><span class="cn">FALSE</span>,nrpoints<span class="op">=</span><span class="fl">10</span>,nbin<span class="op">=</span><span class="fl">500</span>,colramp<span class="op">=</span><span class="va">cols</span><span class="op">)</span></code></pre></div>
<p><img src="dupRadar_files/figure-html/unnamed-chunk-8-1.png" width="1344"></p>
<p>Any further parm sent to the <em>duprateExpPlot</em> is also sent to the <em>smoothScatter</em> function.</p>
</div>
<div id="comparing-the-fitted-parameters-to-other-datasets" class="section level2">
<h2 class="hasAnchor">
<a href="#comparing-the-fitted-parameters-to-other-datasets" class="anchor"></a>Comparing the fitted parameters to other datasets</h2>
<p>The parameters of the fitted model may mean very little (or just nothing) for many, unless there’s other data to compare with. We provide the pre-computed duplication matrices for all the RNA-Seq experiments publicly available from the ENCODE project, for human and mouse.</p>
<div class="figure">
<img src="figures/encode.svg" alt="alt text"><p class="caption">alt text</p>
</div>
<p>With the experience from the ENCODE datasets, we expect from single read experiments little duplication at low RPKM (low <strong>intercept</strong>) rapidly rising because of natural duplication (high <strong>slope</strong>). In contrast, paired-end experiments have a more mild rising of the natural duplication (low <strong>slope</strong>) due to having higher diversity of reads pairs since pairs with same start may still have different end.</p>
<p>The common denominator for both designs is the importance of having a low intercept, suggesting that duplication rate at lowly expressed genes may serve as a quality measure.</p>
<p>All the pre-computed duplication matrices are available in the <a href="https://github.com/ssayols/dupRadar">dupRadar Github site</a>.</p>
</div>
<div id="other-plots" class="section level2">
<h2 class="hasAnchor">
<a href="#other-plots" class="anchor"></a>Other plots</h2>
<p><strong>CAVEAT: Sometimes in discussions duplicate reads (i.e. two physically different reads are mapped to the exact same position) and multi-mapping reads (i.e. a single read is mapped to two or more locations in the genome) are mixed up. <em>dupRadar</em>’s main focus are PCR duplicates in RNA-Seq. However internally we keep track of unique mappers and multimappers, and we use both in some of the examples, to illustrate use cases of our package beyond the main aim. Multi-mapping reads are completely independent of PCR duplicates.</strong></p>
<p>Apart from the plots relating RPK and duplication rate, the <em>dupRadar</em> package provides some other useful plots to extract information about the gene expression levels.</p>
<p>An interesting quality metric are the fraction of reads taken up by groups of genes binned by 5% expression levels.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/readcountExpBoxplot.html">readcountExpBoxplot</a></span><span class="op">(</span>DupMat<span class="op">=</span><span class="va">dm</span><span class="op">)</span></code></pre></div>
<p><img src="dupRadar_files/figure-html/unnamed-chunk-9-1.png" width="672"></p>
<p>In the example we see that the 5% of highest expressed genes in our sample data set take up around 60% of all reads.</p>
<p>The distribution of RPK values per genes can be plotted with:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/expressionHist.html">expressionHist</a></span><span class="op">(</span>DupMat<span class="op">=</span><span class="va">dm</span><span class="op">)</span></code></pre></div>
<p><img src="dupRadar_files/figure-html/unnamed-chunk-10-1.png" width="672"></p>
<p>This would help in identifying skewed distributions with unusual amount of lowly expressed genes, or to detect no consensus between replicates.</p>
</div>
</div>
<div id="other-information-deduced-from-the-data" class="section level1">
<h1 class="hasAnchor">
<a href="#other-information-deduced-from-the-data" class="anchor"></a>Other information deduced from the data</h1>
<p>The duplication rate matrix calculated by the function <em>analyzeDuprates()</em> contains some useful information about the sequencing experiment, that can be used to assess the quality of the data.</p>
<div id="fraction-of-multimappers-per-gene" class="section level2">
<h2 class="hasAnchor">
<a href="#fraction-of-multimappers-per-gene" class="anchor"></a>Fraction of multimappers per gene</h2>
<p>Analogous to per gene duplication rate, the fraction of mutimappers can be easily calculated fom the duplication matrix.</p>
<p>Taking the counts from the column <em>allCountsMulti</em>, and substracting form it the counts from the column <em>allCounts</em>, one can get the total number of multihits. Thus, the fraction of multihits per gene can be calculating then dividing by <em>allCountsMulti</em>.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># calculate the fraction of multimappers per gene</span>
<span class="va">dm</span><span class="op">$</span><span class="va">mhRate</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="va">dm</span><span class="op">$</span><span class="va">allCountsMulti</span> <span class="op">-</span> <span class="va">dm</span><span class="op">$</span><span class="va">allCounts</span><span class="op">)</span> <span class="op">/</span> <span class="va">dm</span><span class="op">$</span><span class="va">allCountsMulti</span>
<span class="co"># how many genes are exclusively covered by multimappers</span>
<span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">dm</span><span class="op">$</span><span class="va">mhRate</span> <span class="op">==</span> <span class="fl">1</span>, na.rm<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 295</code></pre>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># and how many have an RPKM (including multimappers) &gt; 5</span>
<span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">dm</span><span class="op">$</span><span class="va">mhRate</span><span class="op">==</span><span class="fl">1</span> <span class="op">&amp;</span> <span class="va">dm</span><span class="op">$</span><span class="va">RPKMMulti</span> <span class="op">&gt;</span> <span class="fl">5</span>, na.rm<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 8</code></pre>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># and which are they?</span>
<span class="va">dm</span><span class="op">[</span><span class="va">dm</span><span class="op">$</span><span class="va">mhRate</span><span class="op">==</span><span class="fl">1</span> <span class="op">&amp;</span> <span class="va">dm</span><span class="op">$</span><span class="va">RPKMMulti</span> <span class="op">&gt;</span> <span class="fl">5</span>, <span class="st">"ID"</span><span class="op">]</span></code></pre></div>
<pre><code>## [1] GPR89C       LOC728643    LOC606724    TBC1D3C      TBC1D3H     
## [6] MIR650       LOC100133050 HIST1H4J    
## 23228 Levels: 1/2-SBSRNA4 A1BG A1BG-AS1 A1CF A2LD1 A2M A2ML1 A2MP1 ... ZZZ3</code></pre>
<p>We can also generate an overall picture about less extreme cases:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html">hist</a></span><span class="op">(</span><span class="va">dm</span><span class="op">$</span><span class="va">mhRate</span>, 
     breaks<span class="op">=</span><span class="fl">50</span>, 
     col<span class="op">=</span><span class="st">"red"</span>, 
     main<span class="op">=</span><span class="st">"Frequency of multimapping rates"</span>, 
     xlab<span class="op">=</span><span class="st">"Multimapping rate per gene"</span>, 
     ylab<span class="op">=</span><span class="st">"Frequency"</span><span class="op">)</span></code></pre></div>
<p><img src="dupRadar_files/figure-html/unnamed-chunk-12-1.png" width="672"></p>
<p>Also the direct comparison of reads per kilobase between uniquely and multimappers is possible.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># comparison of multi-mapping RPK and uniquely-mapping RPK</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html">plot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span><span class="op">(</span><span class="va">dm</span><span class="op">$</span><span class="va">RPK</span><span class="op">)</span>, 
     <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span><span class="op">(</span><span class="va">dm</span><span class="op">$</span><span class="va">RPKMulti</span><span class="op">)</span>, 
     xlab<span class="op">=</span><span class="st">"Reads per kb (uniquely mapping reads only)"</span>,
     ylab<span class="op">=</span><span class="st">"Reads per kb (all including multimappers, non-weighted)"</span>
<span class="op">)</span></code></pre></div>
<p><img src="dupRadar_files/figure-html/unnamed-chunk-13-1.png" width="672"></p>
<p>Use with identify() to annotate interesting cases interactively.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/identify.html">identify</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span><span class="op">(</span><span class="va">dm</span><span class="op">$</span><span class="va">RPK</span><span class="op">)</span>, 
         <span class="fu"><a href="https://rdrr.io/r/base/Log.html">log2</a></span><span class="op">(</span><span class="va">dm</span><span class="op">$</span><span class="va">RPKMulti</span><span class="op">)</span>,
     labels<span class="op">=</span><span class="va">dm</span><span class="op">$</span><span class="va">ID</span><span class="op">)</span></code></pre></div>
</div>
<div id="connection-between-possible-pcr-artefacts-and-gc-content" class="section level2">
<h2 class="hasAnchor">
<a href="#connection-between-possible-pcr-artefacts-and-gc-content" class="anchor"></a>Connection between possible PCR artefacts and GC content</h2>
<p>In some cases we wondered about influence of GC content on PCR artefacts. An easy way to check using our <em>dupRadar</em> package in conjunction with <em>biomaRt</em> is demonstrated in the following. For simplicity we use our demo data here in which we <em>do not</em> see a big influence.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://www.bioconductor.org/packages/dupRadar">dupRadar</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">biomaRt</span><span class="op">)</span>

<span class="co">## for detailed explanations on biomaRt, please see the respective</span>
<span class="co">## vignette</span>

<span class="co">## set up biomart connection for mouse (needs internet connection)</span>
<span class="va">ensm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/useMart.html">useMart</a></span><span class="op">(</span><span class="st">"ensembl"</span><span class="op">)</span>
<span class="va">ensm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/useDataset.html">useDataset</a></span><span class="op">(</span><span class="st">"mmusculus_gene_ensembl"</span>, mart<span class="op">=</span><span class="va">ensm</span><span class="op">)</span>

<span class="co">## get a table which has the gene GC content for the IDs that have been</span>
<span class="co">## used to generate the table (depends on the GTF annotation that you</span>
<span class="co">## use)</span>
<span class="va">tr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/biomaRt/man/getBM.html">getBM</a></span><span class="op">(</span>attributes<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"mgi_symbol"</span>, <span class="st">"percentage_gc_content"</span><span class="op">)</span>,
            values<span class="op">=</span><span class="cn">TRUE</span>, mart<span class="op">=</span><span class="va">ensm</span><span class="op">)</span>

<span class="co">## create a GC vector with IDs as element names</span>
<span class="va">mgi.gc</span> <span class="op">&lt;-</span> <span class="va">tr</span><span class="op">$</span><span class="va">percentage_gc_content</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">mgi.gc</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">tr</span><span class="op">$</span><span class="va">mgi_symbol</span>

<span class="co">## using dm demo duplication matrix that comes with the package</span>
<span class="co">## add GC content to our demo data and keep only subset for which we can</span>
<span class="co">## retrieve data</span>
<span class="va">keep</span> <span class="op">&lt;-</span> <span class="va">dm</span><span class="op">$</span><span class="va">ID</span> <span class="op">%in%</span> <span class="va">tr</span><span class="op">$</span><span class="va">mgi_symbol</span>
<span class="va">dm.gc</span> <span class="op">&lt;-</span> <span class="va">dm</span><span class="op">[</span><span class="va">keep</span>,<span class="op">]</span>
<span class="va">dm.gc</span><span class="op">$</span><span class="va">gc</span> <span class="op">&lt;-</span> <span class="va">mgi.gc</span><span class="op">[</span><span class="va">dm.gc</span><span class="op">$</span><span class="va">ID</span><span class="op">]</span>

<span class="co">## check distribution of annotated gene GC content (in %)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/boxplot.html">boxplot</a></span><span class="op">(</span><span class="va">dm.gc</span><span class="op">$</span><span class="va">gc</span>, main<span class="op">=</span><span class="st">"Gene GC content"</span>, ylab<span class="op">=</span><span class="st">"% GC"</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="figures/GCcontent.svg" alt="alt text"><p class="caption">alt text</p>
</div>
<p>Now we can compare the dependence of duplication rate on expression level independently for below and above median GC genes (and to mention again, in this data set we don’t have a big difference).</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span><span class="op">)</span>

<span class="co">## below median GC genes</span>
<span class="fu"><a href="../reference/duprateExpDensPlot.html">duprateExpDensPlot</a></span><span class="op">(</span><span class="va">dm.gc</span><span class="op">[</span><span class="va">dm.gc</span><span class="op">$</span><span class="va">gc</span><span class="op">&lt;=</span><span class="fl">45</span>,<span class="op">]</span>, main<span class="op">=</span><span class="st">"below median GC genes"</span><span class="op">)</span>

<span class="co">## above median GC genes</span>
<span class="fu"><a href="../reference/duprateExpDensPlot.html">duprateExpDensPlot</a></span><span class="op">(</span><span class="va">dm.gc</span><span class="op">[</span><span class="va">dm.gc</span><span class="op">$</span><span class="va">gc</span><span class="op">&gt;=</span><span class="fl">45</span>,<span class="op">]</span>, main<span class="op">=</span><span class="st">"above median GC genes"</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="figures/GCdependence.png" alt="alt text"><p class="caption">alt text</p>
</div>
</div>
</div>
<div id="conclusion" class="section level1">
<h1 class="hasAnchor">
<a href="#conclusion" class="anchor"></a>Conclusion</h1>
<p>The <em>dupRadar</em> package provides a framework for the analysis of duplicate rates in RNAseq datasets. In addition, it includes a set of convenient wrappers to correctly call some common tools used for marking PCR duplicated reads in the BAM file. It’s shipped as Bioconductor package in order to offer a common framework for all the tools involved, and simplify its use.</p>
</div>
<div id="including-dupradar-into-pipelines" class="section level1">
<h1 class="hasAnchor">
<a href="#including-dupradar-into-pipelines" class="anchor"></a>Including dupRadar into pipelines</h1>
<p>To include dupRadar as a single step in an RNA-Seq pipeline, integration into a short R-script can be done like in the following:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#!/usr/bin/env Rscript</span>

<span class="co">########################################</span>
<span class="co">##</span>
<span class="co">## dupRadar shell script</span>
<span class="co">## call dupRadar R package from the shell for </span>
<span class="co">## easy integration into pipelines</span>
<span class="co">##</span>
<span class="co">## Holger Klein &amp; Sergi Sayols</span>
<span class="co">##</span>
<span class="co">## https://github.com/ssayols/dupRadar</span>
<span class="co">##</span>
<span class="co">## input:</span>
<span class="co">## - _duplicate marked_ bam file</span>
<span class="co">## - gtf file</span>
<span class="co">## - parameters for duplication counting routine:</span>
<span class="co">##   stranded, paired, outdir, threads.</span>
<span class="co">##</span>
<span class="co">########################################</span>

<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://www.bioconductor.org/packages/dupRadar">dupRadar</a></span><span class="op">)</span>

<span class="co">####################</span>
<span class="co">##</span>
<span class="co">## get name patterns from command line</span>
<span class="co">##</span>
<span class="va">args</span>   <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/commandArgs.html">commandArgs</a></span><span class="op">(</span><span class="cn">TRUE</span><span class="op">)</span>

<span class="co">## the bam file to analyse</span>
<span class="va">bam</span> <span class="op">&lt;-</span> <span class="va">args</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>
<span class="co">## usually, same GTF file as used in htseq-count</span>
<span class="va">gtf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span><span class="st">"gtf="</span>,<span class="st">""</span>,<span class="va">args</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span>
<span class="co">## no|yes|reverse</span>
<span class="va">stranded</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span><span class="st">"stranded="</span>,<span class="st">""</span>,<span class="va">args</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span><span class="op">)</span>
<span class="co">## is a paired end experiment</span>
<span class="va">paired</span>   <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span><span class="st">"paired="</span>,<span class="st">""</span>,<span class="va">args</span><span class="op">[</span><span class="fl">4</span><span class="op">]</span><span class="op">)</span>
<span class="co">## output directory</span>
<span class="va">outdir</span>   <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span><span class="st">"outdir="</span>,<span class="st">""</span>,<span class="va">args</span><span class="op">[</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span>
<span class="co">## number of threads to be used</span>
<span class="va">threads</span>  <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/integer.html">as.integer</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span><span class="st">"threads="</span>,<span class="st">""</span>,<span class="va">args</span><span class="op">[</span><span class="fl">6</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>

<span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">args</span><span class="op">)</span> <span class="op">!=</span> <span class="fl">6</span><span class="op">)</span> <span class="op">{</span> 
  <span class="kw"><a href="https://rdrr.io/r/base/stop.html">stop</a></span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"Usage: ./dupRadar.sh &lt;file.bam&gt; &lt;genes.gtf&gt; "</span>,
               <span class="st">"&lt;stranded=[no|yes|reverse]&gt; paired=[yes|no] "</span>,
               <span class="st">"outdir=./ threads=1"</span><span class="op">)</span><span class="op">)</span>
<span class="op">}</span>

<span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files.html">file.exists</a></span><span class="op">(</span><span class="va">bam</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span> 
  <span class="kw"><a href="https://rdrr.io/r/base/stop.html">stop</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"File"</span>,<span class="va">bam</span>,<span class="st">"does NOT exist"</span><span class="op">)</span><span class="op">)</span>
<span class="op">}</span>

<span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files.html">file.exists</a></span><span class="op">(</span><span class="va">gtf</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
  <span class="kw"><a href="https://rdrr.io/r/base/stop.html">stop</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"File"</span>,<span class="va">gtf</span>,<span class="st">"does NOT exist"</span><span class="op">)</span><span class="op">)</span>
<span class="op">}</span>

<span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files.html">file.exists</a></span><span class="op">(</span><span class="va">outdir</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span> 
  <span class="kw"><a href="https://rdrr.io/r/base/stop.html">stop</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"Dir"</span>,<span class="va">outdir</span>,<span class="st">"does NOT exist"</span><span class="op">)</span><span class="op">)</span>
<span class="op">}</span>

<span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">stranded</span><span class="op">)</span> <span class="op">|</span> <span class="op">!</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">grepl</a></span><span class="op">(</span><span class="st">"no|yes|reverse"</span>,<span class="va">stranded</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
  <span class="kw"><a href="https://rdrr.io/r/base/stop.html">stop</a></span><span class="op">(</span><span class="st">"Stranded has to be no|yes|reverse"</span><span class="op">)</span>
<span class="op">}</span>

<span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">paired</span><span class="op">)</span> <span class="op">|</span> <span class="op">!</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">grepl</a></span><span class="op">(</span><span class="st">"no|yes"</span>,<span class="va">paired</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
  <span class="kw"><a href="https://rdrr.io/r/base/stop.html">stop</a></span><span class="op">(</span><span class="st">"Paired has to be no|yes"</span><span class="op">)</span>
<span class="op">}</span>

<span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">threads</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
  <span class="kw"><a href="https://rdrr.io/r/base/stop.html">stop</a></span><span class="op">(</span><span class="st">"Threads has to be an integer number"</span><span class="op">)</span>
<span class="op">}</span>

<span class="va">stranded</span> <span class="op">&lt;-</span> <span class="kw">if</span><span class="op">(</span><span class="va">stranded</span> <span class="op">==</span> <span class="st">"no"</span><span class="op">)</span> <span class="fl">0</span> <span class="kw">else</span> <span class="kw">if</span><span class="op">(</span><span class="va">stranded</span> <span class="op">==</span> <span class="st">"yes"</span><span class="op">)</span> <span class="fl">1</span> <span class="kw">else</span> <span class="fl">2</span>

<span class="co">## end command line parsing</span>
<span class="co">##</span>
<span class="co">########################################</span>

<span class="co">########################################</span>
<span class="co">##</span>
<span class="co">## analyze duprates and create plots</span>
<span class="co">##</span>
<span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="st">"Processing file "</span>, <span class="va">bam</span>, <span class="st">" with GTF "</span>, <span class="va">gtf</span>, <span class="st">"\n"</span><span class="op">)</span>

<span class="co">## calculate duplication rate matrix</span>
<span class="va">dm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/analyzeDuprates.html">analyzeDuprates</a></span><span class="op">(</span><span class="va">bam</span>,
                      <span class="va">gtf</span>,
                      <span class="va">stranded</span>,
                      <span class="op">(</span><span class="va">paired</span> <span class="op">==</span> <span class="st">"yes"</span><span class="op">)</span>,
                      <span class="va">threads</span><span class="op">)</span>

<span class="co">## produce plots</span>

<span class="co">## duprate vs. expression smooth scatter</span>
<span class="fu"><a href="https://rdrr.io/r/grDevices/png.html">png</a></span><span class="op">(</span>file<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">outdir</span>,<span class="st">"/"</span>,<span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span><span class="st">"(.*)\\.[^.]+"</span>,<span class="st">"\\1"</span>,<span class="fu"><a href="https://rdrr.io/r/base/basename.html">basename</a></span><span class="op">(</span><span class="va">bam</span><span class="op">)</span><span class="op">)</span>,<span class="st">"_dupRadar_drescatter.png"</span><span class="op">)</span>,
    width<span class="op">=</span><span class="fl">1000</span>, height<span class="op">=</span><span class="fl">1000</span><span class="op">)</span>
<span class="fu"><a href="../reference/duprateExpDensPlot.html">duprateExpDensPlot</a></span><span class="op">(</span><span class="va">dm</span>, main<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/basename.html">basename</a></span><span class="op">(</span><span class="va">bam</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html">dev.off</a></span><span class="op">(</span><span class="op">)</span>

<span class="co">## expression histogram</span>
<span class="fu"><a href="https://rdrr.io/r/grDevices/png.html">png</a></span><span class="op">(</span>file<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">outdir</span>,<span class="st">"/"</span>,<span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span><span class="st">"(.*)\\.[^.]+"</span>,<span class="st">"\\1"</span>,<span class="fu"><a href="https://rdrr.io/r/base/basename.html">basename</a></span><span class="op">(</span><span class="va">bam</span><span class="op">)</span><span class="op">)</span>,<span class="st">"_dupRadar_ehist.png"</span><span class="op">)</span>,
    width<span class="op">=</span><span class="fl">1000</span>, height<span class="op">=</span><span class="fl">1000</span><span class="op">)</span>
<span class="fu"><a href="../reference/expressionHist.html">expressionHist</a></span><span class="op">(</span><span class="va">dm</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html">dev.off</a></span><span class="op">(</span><span class="op">)</span>

<span class="co">## duprate vs. expression boxplot</span>
<span class="fu"><a href="https://rdrr.io/r/grDevices/png.html">png</a></span><span class="op">(</span>file<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">outdir</span>,<span class="st">"/"</span>,<span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span><span class="st">"(.*)\\.[^.]+"</span>,<span class="st">"\\1"</span>,<span class="fu"><a href="https://rdrr.io/r/base/basename.html">basename</a></span><span class="op">(</span><span class="va">bam</span><span class="op">)</span><span class="op">)</span>,<span class="st">"_dupRadar_drebp.png"</span><span class="op">)</span>,
    width<span class="op">=</span><span class="fl">1000</span>, height<span class="op">=</span><span class="fl">1000</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mar<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">10</span>,<span class="fl">4</span>,<span class="fl">4</span>,<span class="fl">2</span><span class="op">)</span><span class="op">+</span><span class="fl">.1</span><span class="op">)</span>
<span class="fu"><a href="../reference/duprateExpBoxplot.html">duprateExpBoxplot</a></span><span class="op">(</span><span class="va">dm</span>, main<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/basename.html">basename</a></span><span class="op">(</span><span class="va">bam</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html">dev.off</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<div id="citing-dupradar" class="section level2">
<h2 class="hasAnchor">
<a href="#citing-dupradar" class="anchor"></a>Citing dupRadar</h2>
<p>Please consider citing dupRadar if used in support of your own research:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/citation.html">citation</a></span><span class="op">(</span><span class="st">"dupRadar"</span><span class="op">)</span></code></pre></div>
<pre><code>## 
##   Sergi Sayols, Denise Scherzinger and Holger Klein (2016): dupRadar: a
##   Bioconductor package for the assessment of PCR artifacts in RNA-Seq
##   data. BMC Bioinformatics, 17:428, doi:10.1186/s12859-016-1276-2
## 
## A BibTeX entry for LaTeX users is
## 
##   @Article{,
##     title = {dupRadar: a Bioconductor package for the assessment of PCR artifacts in RNA-Seq data},
##     author = {Sergi Sayols and Denise Scherzinger and Holger Klein},
##     year = {2016},
##     journal = {BMC Bioinformatics},
##     doi = {10.1186/s12859-016-1276-2},
##     url = {http://dx.doi.org/10.1186/s12859-016-1276-2},
##     volume = {17},
##     issue = {1},
##     pages = {428},
##   }</code></pre>
</div>
<div id="reporting-problems-or-bugs" class="section level2">
<h2 class="hasAnchor">
<a href="#reporting-problems-or-bugs" class="anchor"></a>Reporting problems or bugs</h2>
<p>If you run into problems using dupRadar, the <a href="https://support.bioconductor.org/">Bioconductor Support site</a> is a good first place to ask for help. If you think there is a bug or an unreported feature, you can report it using the <a href="https://github.com/ssayols/dupRadar">dupRadar Github site</a>.</p>
</div>
</div>
<div id="session-info" class="section level1">
<h1 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session info</h1>
<p>The following package and versions were used in the production of this vignette.</p>
<pre><code>## R version 4.0.3 (2020-10-10)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 20.04.2 LTS
## 
## Matrix products: default
## BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.9.0
## LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.9.0
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] dupRadar_1.16.1  knitr_1.30       BiocStyle_2.16.1
## 
## loaded via a namespace (and not attached):
##  [1] magrittr_2.0.1      lattice_0.20-41     R6_2.5.0           
##  [4] ragg_0.4.0          rlang_0.4.9         stringr_1.4.0      
##  [7] tools_4.0.3         grid_4.0.3          xfun_0.19          
## [10] KernSmooth_2.23-18  htmltools_0.5.0     systemfonts_0.3.2  
## [13] yaml_2.2.1          assertthat_0.2.1    rprojroot_2.0.2    
## [16] digest_0.6.27       Rsubread_2.2.6      pkgdown_1.6.1      
## [19] crayon_1.3.4        bookdown_0.21       textshaping_0.2.1  
## [22] Matrix_1.3-2        BiocManager_1.30.10 fs_1.5.0           
## [25] codetools_0.2-18    memoise_1.1.0       evaluate_0.14      
## [28] rmarkdown_2.5       stringi_1.5.3       compiler_4.0.3     
## [31] desc_1.2.0</code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Sergi Sayols.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
